experimental/devices/mfrc522/rf522.go

```diff
@@ -0,0 +1,763 @@
+// Copyright 2016 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package mfrc522 controls a Mifare RFID card reader.
+// The code is largely ported from Python library : https://github.com/mxgxw/MFRC522-python
+// The datasheet is available at : https://www.nxp.com/docs/en/data-sheet/MFRC522.pdf
+package mfrc522
+
+import (
+	"errors"
+	"fmt"
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/gpio/gpioreg"
+	"periph.io/x/periph/conn/spi"
+	"periph.io/x/periph/experimental/devices/mfrc522/commands"
+	"time"
+)
+
+// Dev defines the RFID reader properties, the SPI bus and GPIO pins.
+type Dev struct {
+	ResetPin      gpio.PinIO
+	IrqPin        gpio.PinIO
+	Authenticated bool
+	antennaGain   int
+	MaxSpeedHz    int
+	spiDev        spi.Conn
+}
+
+// DefaultKey provides the default bytes for card authentication for method B.
+var DefaultKey = []byte{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+
+// NewSPI creates and initializes the RFID card reader attached to SPI.
+// - spiPort - the SPI device to use.
+// - maxSpeed - maximum speed for SPI data transfers.
+// - resetPin - reset GPIO pin.
+// - irqPin - irq GPIO pin.
+func NewSPI(spiPort spi.Port, maxSpeed int, resetPin, irqPin string) (device *Dev, err error) {
+
+	spiDev, err := spiPort.Connect(int64(maxSpeed), spi.Mode0, 8)
+	if err != nil {
+		return
+	}
+
+	if err != nil {
+		return
+	}
+
+	dev := &Dev{
+		spiDev:     spiDev,
+		MaxSpeedHz: maxSpeed,
+	}
+
+	pin := gpioreg.ByName(resetPin)
+	dev.ResetPin = pin
+	dev.ResetPin.Out(gpio.High)
+
+	pin = gpioreg.ByName(irqPin)
+	dev.IrqPin = pin
+	dev.IrqPin.In(gpio.PullUp, gpio.FallingEdge)
+
+	err = dev.Init()
+
+	device = dev
+
+	return
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() (err error) {
+	err = r.Reset()
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TModeReg, 0x8D)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TPrescalerReg, 0x3E)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TReloadRegL, 30)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TReloadRegH, 0)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.TxAutoReg, 0x40)
+	if err != nil {
+		return
+	}
+	err = r.devWrite(commands.ModeReg, 0x3D)
+	if err != nil {
+		return
+	}
+	err = r.SetAntenna(true)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (r *Dev) writeSpiData(dataIn []byte) (out []byte, err error) {
```

Comment:
Inline this function.

Reply:
It is used in `devRead` and `devWrite`, does it make sense to copy/paste it there?

Reply:
devWrite doesn't use out, so it boils down to the direct Tx call.

Reply:
Any call to RFID returns something, and it's full-duplex. So the buffer to receive the result should be of the same size as the input buffer. Even if I don't use the result - I still need to pass the buffer to fill in.

Reply:
You don't. The bytes read will be discarded. Here's an example: https://github.com/google/periph/blob/master/host/sysfs/spi.go#L364